ipsec, auth: Avoid out-of-order events that lead to inconsistent node IDs

pchaigno opened the pull request:
We found out that the logic to allocate and deallocate node IDs had multiple bugs because it wasn't handling out-of-order events properly. In this context, out-of-order events are e.g. when pod objects are received before their corresponding node object. We discussed those issues at length in https://github.com/cilium/cilium/pull/26725 (see review) and concluded we should try to avoid having to deal with such out-of-order events altogether.

This pull request therefore removes all handling of node IDs on pod events. Node IDs are now entirely managed from node events. The node IDs are removed from the ipcache (`ep->node_id`). That simplifies the agent code a lot (all ipcache is uncoupled from node ID logic) and only slightly complicated the datapath logic (one more map lookup).

Summary of commits:
- Commits 1-4 are preparatory work in the datapath and the agent, without functional changes.
- Commit 5 moves the node ID allocation to be consistent with the deallocation.
- Commit 6 introduces a new DROP reason.
- Commits 7-8 remove the two consumers of `ep->node_id` in the datapath and replace them with an additional map lookup.
- Commits 9-10 covers the new logic of the previous commit in BPF unit tests.
- Commits 11-12 remove the node IDs from the ipcache and clean up related agent logic.
- Commit 13 fixes a leak of node IDs when the IP addresses of a node are updated.
- Commit 14 adds a runtime check for the bogus state fixed in the previous commit.
- Commits 15-16 cover the node ID lifecycle in Golang integration tests.

See commits for details.

This PR is an alternative to the solution tried at https://github.com/cilium/cilium/pull/26725.

```release-note
Fix a bug that could cause packet drops of type XfrmOutPolBlock when IPsec is enabled and node are recycled.
Fix a bug that could cause IPsec-encrypted packets to be sent to the wrong destination node when node churn is high.
```

pchaigno committed:
monitor: New NO_NODE_ID drop type

This new packet drop type/reason will be used by IPsec and Mutual Auth
when a node ID lookup doesn't return anything. Both features rely on the
presence of the node ID to function properly.

A failure to find the node ID can occur if traffic is sent to a pod on
a node for which we didn't receive the NodeAdd or NodeUpdate event yet.
Such out-of-order events (PodAdd received before NodeAdd) are possible
and normal behavior.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf/auth: Stop relying on ipcache node_id field

The Mutual Auth feature keeps track of authenticated remote nodes in a
BPF map keyed by security identities and by the remote node ID. It
currently retrieves that remote node ID from the ipcache.

In a subsequent commit, however, we will remove this node ID information
from the ipcache (to simplify node ID management in the agent). So the
Mutual Auth datapath must retrieve the node ID from a different place.

Fortunately, we already have a BPF map containing all node IP to node ID
mappings. Thus, in this commit, we use the remote node IP obtained from
the ipcache (ep->tunnel_endpoint) to lookup the remote node ID in this
BPF map. That additional map lookup removes the dependency of Mutual
Auth on the node_id field of the ipcache.

An alternative approach would be to modify the auth BPF map to be keyed
by remote node IP instead of remote node ID. That is however a much
bigger change, both in terms of agent changes and potential impact.

Note we need to handle a special case here for when the remote node ID
(ep->tunnel_endpoint) is 0.0.0.0. That corresponds to the local node, so
ipcache entries of local pods. To maintain the previous mutual auth
behavior where an entry was created in the auth map for the local node,
we set the remote_node_id to 0 in that case. In userspace, 0 is also the
node ID used for the local node (although it's not populated in the node
ID map).

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf/Makefile: Remove IPsec from compile options for bpf_{xdp,sock}

IPsec isn't used or compatible with bpf_xdp and bpf_sock, so there is
no reason to compile test it there. This will simplify subsequent
commits where we will introduce some code in IPsec that isn't compatible
with non-tc BPF hooks.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf: Move IPsec prep logic to encrypt.h

This commit has no functional changes. It simply moves the logic to
prepare packets for IPsec encryption to the dedicated encrypt.h file.
Function set_ipsec_encrypt_mark is also created to simplify a subsequent
commit (no, it won't remain a function that simply calls another
function :)).

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf/ipsec: Stop relying on ipcache node_id field

The IPsec feature relies on the node IDs encoded in the packet marks to
match packets against encryption rules (XFRM OUT policies and states).
It currently retrieves the ID of the remote node from the ipcache.

In a subsequent commit, however, we will remove this node ID information
from the ipcache (to simplify node ID management in the agent). So
IPsec's datapath must retrieve the node ID from a different place.

Fortunately, we already have a BPF map containing all node IP to node ID
mappings. Thus, in this commit, we use the remote node IP obtained from
the ipcache (ep->tunnel_endpoint) to lookup the remote node ID in this
BPF map.

This one additional hashmap lookup is expected to have a negligible cost
compared to the cost of encryption. It should even be small compared to
the cost of the ipcache LPM lookup.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
ipcache: Stop populating node IDs in ipcache

In previous commits, we removed all consumers (IPsec and Mutual Auth)
of the ipcache's node_id field. This commit removes population of those
node IDs in the ipcache.

**Problem**

Needing to populate the ipcache with node IDs causes all sorts of
issues. Node IDs are allocated for and mapped with specific node IP
addresses. In practice, it means node IDs are tied to the lifecycle of
both node and pod objects:
1. We need to allocate node IDs when receiving node objects with new
   node IP addresses (either new node objects or existing node objects
   with changing IP addresses).
2. We need to allocate node IDs when receiving a node IP address via a
   pod (hostIP of the pod).

Node objects have all node IP addresses. So why are we not relying
solely on 1? We also need 2 because pod adds and pod updates (with new
hostIP) can arrive before the related node add or node update.
Basically, kube-apiserver doesn't provide any guarantee of ordering
between node and pod events. Since we need to populate the ipcache with
a node ID for each remote pod, we need to allocate a node ID when we
receive the pod object and not later, when we receive the node object.

Unfortunately, tying the node IDs to two object lifecycles in this way
causes a lot of complications. For example, we can't simply deallocate
node IDs on pod deletions because other pods may still be using the node
ID. We also can't deallocate simply on node deletions because pods may
still be using it. The typical solution to such issues is to maintain a
reference count, but that leads to its own set of issues in this case
[1].

**Considered Solution**

An alternative to all this is to force events to always be managed in
the same order. For example, NodeAdd -> PodAdd -> PodDelete ->
NodeDelete. The agent unfortunately doesn't easily lend itself to that.
Without a major refactoring, we would have to populate the ipcache as
usual, but skip node IDs. Then, on NodeAdd, we would allocate a node ID
and go through the ipcache again to populate the node IDs we skipped.
In the datapath, we would have an explicit drop for such half-populated
ipcache entries.

To implement this without needing to walk through the entire ipcache for
each NodeAdd or NodeUpdate, we need to keep track of the specific
entries that need to be populated. We need the same sort of mechanism
for deletions. This solution quickly turns into something quite complex
to implement with good performance.

**This Solution**

It should be quite clear at this point that this problem's complexity
stems from having node IDs tied to both the pods and nodes' lifecycles.

To untie node IDs from pod lifecycles, and thus, stop populating the
ipcache with node IDs, we will need the datapath to retrieve them from
some other place. Fortunately, we already have a BPF map that maps node
IP addresses to node IDs. It is currently only used as a restoration
mechanism for agent restarts, but there's no reason we can't use it for
more.

Thus, with the node IP address retrieved from the ipcache
(ep->tunnel_endpoint), we can lookup the node ID in the BPF map.

Of course, since pod objects can arrive before their corresponding node
object, the node ID may not always be ready in the BPF map when we need
it. When that happens, we will drop the packet with a specific drop code
(DROP_NO_NODE_ID). These drops simply mean we tried to send traffic to a
node for which we didn't receive the node object yet (hence the
encryption configuration isn't setup yet).

These datapath changes have been implemented in the previous commit.
This commit removes all population of node IDs in the ipcache.

We can also remove node IDs from the tunnel map. It matters less since
the tunnel map is tied to node objects, but let's keep the datapath
consistent between the two maps (ipcache and tunnel map).

1 - https://github.com/cilium/cilium/pull/26725#discussion_r1266785189
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
ipsec: Refactor IPsec logic to allocate node IDs in one place

This commit has no functional changes. It simply moves the allocation of
node IDs to a single place in the top calling function.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

jrfastab reviewed (approved):
This is amazing!

pchaigno committed:
datapath, ipcache: Remove unused node ID functions

The AllocateNodeID function was used to allocate node IDs from the
ipcache logic. Since we don't populate the ipcache with node IDs
anymore, it's unused and can be removed. The DeallocateNodeID function
was never used (a sibling function was used instead).

This commit has no functional changes.

Reported-by: John Fastabend <john.fastabend@gmail.com>
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

jrfastab commented on pkg/ipcache/ipcache.go:
```diff
@@ -653,14 +643,10 @@ func (ipc *IPCache) deleteLocked(ip string, source source.Source) (namedPortsCha
 	}
 
-	if newHostIP != nil {
-		nodeID = ipc.AllocateNodeID(newHostIP)
```
by the way I suspect this allocateNodeID from the delete path is the bug most often hit. Shouldn't have this been a deleteNodeID? Or rather don't we need some delete logic here as added in the linked PR. Anyways now we don't need to reason about it.

pchaigno replied:
I pushed your commit (with a new commit description) to fix this.

jrfastab commented on pkg/maps/ipcache/ipcache.go:
```diff
@@ -124,14 +123,13 @@ func NewKey(ip net.IP, mask net.IPMask, clusterID uint8) Key {
 type RemoteEndpointInfo struct {
 	SecurityIdentity uint32     `align:"sec_identity"`
 	TunnelEndpoint   types.IPv4 `align:"tunnel_endpoint"`
-	NodeID           uint16     `align:"node_id"`
```
Do we need to patch the hole here with `Pad uint16` then when upgrade/downgrade we don't have to adjust the maps formats? Seems this just pushes the pad to the end.

pchaigno committed:
bpf/tests: Cover DROP_NO_NODE_ID case

If the node ID for the destination node of a pod-to-pod packet is not
found in the node ID map, we drop the packet with DROP_NO_NODE_ID. This
can happen when the k8s Node object for that destination hasn't been
received yet. This commit covers that case in BPF unit tests.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf/tests: Rename IPsec unit tests

This will be useful to enforce ordering between different tests, with
the first one preparing part of the maps for subsequent tests. Numbering
starts at 2 because the next commit will introduce a test 1.

"ipsec_" is removed from the test names to shorten it a bit (the IPsec
context should be clear from the test filename).

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
node: Fix leak of node IP <> ID mappings

This commit fixes a leak of node IP to node ID mappings when the set of
IP addresses of a node is updated. We have logic to allocate and map
node IDs on node creation, logic to unmap and deallocate node IDs on
node deletions, but we didn't fully handle the case of node updates.

On node updates, we would map new IP addresses to the already-allocated
node ID. We forgot to unmap the old IP addresses. This commit fixes that
oversight.

This leak means that some IP addresses that were previously assigned to
the node would remain mapped to the node ID. Those IP addresses could
then be assigned to a new node. The new node would then receive the same
node ID as the old node (because we first check if any of the node IPs
already have an ID). This could lead to encrypted traffic being sent to
the wrong node.

Before removing the node IDs from the ipcache (see previous commit),
this leak could have another consequence. If after the node ID had been
reassigned to the new node, the old node was deleted, the node ID would
be deallocated, but it would remain in the ipcache. The XFRM policies
and states of the node would also be identified based on the node ID and
removed. Thus, both the old and new nodes' XFRM configs would be
removed. As a consequence, outgoing packets for those nodes would be
dropped with XfrmOutPolBlock (the default drop-all rule).

Fixes: af88b42bd ("datapath: Introduce node IDs")
Signed-off-by: John Fastabend <john.fastabend@gmail.com>
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
node_ids: Check that two nodes don't share a node ID

The previous commit fixed a bug that could lead to two nodes sharing the
same node ID. This commit adds a check, on node deletion, for such
erroneous state. It completes the existing check that a single node
doesn't have multiple node IDs. In both cases, an error is thrown such
that users will notice and our CI will fail.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
nodemap: Extend fake node map for use in tests

The fake node ID map currently only defines the minimum set of functions
needed for the interface. This commit extends it to simulate the actual
behavior of the BPF map with an in-memory Golang map.

This will allow us to use check that the map content matches our
expectations for various events of the node object lifecycle.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
node: Allocate node ID even if IPsec is disabled

Before this commit, we would allocate a node ID for new nodes if IPsec
was enabled. That wasn't consistent with (1) allocations triggered by
pod creations, where we allocate even if IPsec is disabled, and (2)
deallocations triggered by node deletions which also happen regardless
of IPsec being enabled.

This commit changes the nodeUpdate logic slightly such that the node ID
is allocated regardless of IPsec being enabled. As a consequence, we
need to update several integration tests to mock the node ID BPF map.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
node: Integration test for node ID lifecycle

The node ID lifecycle is tied to the node lifecycle. This commit adds an
integration test to cover this lifecycle. It covers the creation and
deletion of nodes, as well as the update of a node's IP addresses. It
also checks that two nodes don't end up with the same node ID and that
each node has a single node ID for all its IP addresses.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
nodemap: Run tests on dedicated map

We typically don't execute privileged integration tests on the usual BPF
maps, but instead use copies prefixed with "test_". Let's do the same
for the node ID map tests.

Fixes: ef26119ba72 ("nodeid map: provide map via hive cell")
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

gandro commented on pkg/ipcache/types/types.go:
```diff
@@ -63,7 +63,6 @@ func NewResourceID(kind ResourceKind, namespace, name string) ResourceID {
 
 // NodeIDHandler is responsible for the management of node identities.
 type NodeIDHandler interface {
```
This IPCache type is no longer needed, is it? Can we remove it and remove it from the IPCache `Configuration` completely? Thanks

pchaigno replied:
The Mutual Auth feature still uses it:
```
$ git grep -P "ipCache.GetNodeI[PD]" pkg/auth/
pkg/auth/authmap_gc.go:                 nodeID, exists := r.ipCache.GetNodeID(addr.IP)
pkg/auth/manager.go:    nodeIP := a.ipCache.GetNodeIP(key.remoteNodeID)
```
I suspect that could be avoided by giving the auth package a reference to the node manager directly, but I'm a bit hesitant to make such changes to a feature I have very little context on. I've pinged Mutual Auth folks about this PR and Marco will review, so I'm hoping they can carry additional simplifications as follow ups.

mhofstetter replied:
Yep, should be easy to get rid of this and depend directly on the nodemanager. Will come up with a follow up.

gandro replied:
Ah, I see. Thanks!

mhofstetter replied:
@gandro @pchaigno see https://github.com/cilium/cilium/pull/27106

gandro commented on pkg/datapath/linux/node_ids.go:
```diff
@@ -187,6 +187,25 @@ func (n *linuxNodeHandler) unmapNodeID(ip string) error {
+nextOldIP:
+	for _, oldAddr := range oldIPs {
+		for _, newAddr := range newIPs {
+			if newAddr.IP.String() == oldAddr.IP.String() {
```
Any reason we are not using `net.IP.Equal()` here?

pchaigno replied:
None. I pushed an update with only rebase + that change.

gandro reviewed (commented):
My code owners mostly look excellent to me, thanks a lot for reducing the complexity in the agent code

I had two minor points that stood out to me

mhofstetter commented on bpf/lib/auth.h:
```diff
@@ -10,17 +10,30 @@
+	__u16 *node_id;
+
+	if (remote_node_ip) {
+		node_ip.family = ENDPOINT_KEY_IPV4;
```
just for clarification: the tunnel IP is always IPv4?

pchaigno replied:
Yep, that's correct.

mhofstetter reviewed (approved):
proxy & servicemesh related bits LGTM! thanks a lot @pchaigno
